### PR TITLE
suppress browser prompt on authenticate failure

### PR DESCRIPTION
**Pull request by machisuji**
If anyone knows a better way to do this let me know. The motivation is that a user should not get to see a password prompt in the browser when their session expired while doing stuff in the work packages list.

WP [#20593](https://community.openproject.org/work_packages/20593)

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://machisuji-feature-suppress-basic-auth-prompt.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://machisuji-feature-suppress-basic-auth-prompt.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: spec/requests/api/v3/authentication_spec.rb [hunk 1]
```diff
@@ -66,7 +66,7 @@ def basic_auth(user, password)
         end
 
         it 'should return the WWW-Authenticate header' do
-          expect(response.header['WWW-Authenticate']).to include 'Basic realm="OpenProject"'
+          expect(response.header['WWW-Authenticate']).to include 'BasicAuth realm="OpenProject API"'
```
Comment (NobodysNightmare) [inline spec/requests/api/v3/authentication_spec.rb]: @houndci someone should tell you that 100 is the real limit...

**Comment by oliverguenther** [comment]
While I concur with the reasoning behind this change, it raises a more profound question I did not see in the basic auth PR. 

Why exactly do we output 401 WWW-Authenticate in this case?
For session-based API authentication, wouldn't the proper response be `403 Forbidden` when the session is invalid? There is nothing the user can do (especially if authentication was performed on an external service with omniauth).

**Comment by machisuji** [comment]
We return the header because that's what the spec says. It doesn't say we have to return the Basic scheme in the header, though. I mean what we could do is only return the Basic scheme in the header when the user actually tried to authenticate via basic auth.

Otherwise we could return another challenge, i.e. 'Cookie' (which refers to the session-based authentication). Meaning that with the right cookie (session) you may have access (when you're logged in in that session).

**Comment by machisuji** [comment]
Something in that vein is done [here](https://github.com/opf/openproject/blob/dev/app/controllers/application_controller.rb#L259-L263). Not quite sure about the request header and who sends that when. But independently from that we could just send 'Session' or 'Cookie' or what ever you might want to call it as the challenge by default.

The spec allows for several schemes. As I interpret it the `WWW-Authenticate` header should indicate possible ways to authenticate when returning a 401. So technically one could return several schemes: Cookie, Basic and (later) OAuth.

But this topic is very fuzzy and I'm not even sure how well this is supported by the clients to have several schemes.

**Comment by oliverguenther** [comment]
Okay, bear with me: Trying to parse your comments and recap into a possible path.

OP could communicate through the `WWW-Authenticate` header which authentication methods are available. Browsers officially support Basic (and other options such as Digest, etc., none of which are of interest currently), but according to RFC 2617:

`401 Unauthorized`

> ... MUST include a WWW-Authenticate header field containing **at least one challenge** 
> applicable to the requested resource

Thus I would assume that clients can handle (and ignore) unknown values in the header. If so, that would be a suitable method of communicating further options, possibly added by plugins, such as OAuth, etc.

A `X-Authentication-Scheme = Session` header is currently [appended in the Frontend for all ajax calls](https://github.com/opf/openproject/blob/dev/app/assets/javascripts/application.js.erb#L539-L551).

This header could be used to force an authentication scheme used for an API call. From this value, the optimal authentication flow can be determined.

### Examples

| `X-Authentication-Scheme` value | Authentication successful if ... | Authentication |
| --- | --- | --- |
| (empty) | cf. Basic | cf. Basic |
| Basic | Credentials of `WWW-Authenticate` valid according to PR 2978 | Respond `401 Unauthorized` with `WWW-Authenticate: Basic ..` |
| Session | Session exists and is valid | Respond `403 Forbidden` unless Session exists |
| OAuth | Validated url token or `Authorization: Bearer ...` (depending on the actual implementation) | Redirect to OAuth service endpoint / return service url for desktop flow |

### Implementation

With Warden, it shouldn't be too difficult to express the different schemes. If I recall correctly, currently all strategies are tried in order they are defined in the initializer. With more schemes added to OpenProject, this behavior needs to change anyway.

This implementation would provide a handler for the section you linked from ApplicationController, so instead that handler is called to determine the authentication flow for the specified `X-Authentication-Scheme`. This should be extracted from the ApplicationController, anyway.

**Comment by NobodysNightmare** [comment]
@machisuji I just realized that you need to base this work on `release/4.2`. If you already did that it is fine, the merger can then

:warning: **merge this branch into `release/4.2` first.**

If not you should rebase your branch onto `release/4.2`, to allow merging into `release/4.2`.

On the technical side of the discussion I will for now leave you alone with @oliverguenther, as I trust him to know more about the matter than I do :)

**Comment by machisuji** [comment]
@NobodysNightmare Will base it on release/4.2.

@oliverguenther I agree. I will adapt it to consider the X-Authentication-Scheme header. Though I disagree on the 403 for the session scheme. IMO it should still be 401 if you're not logged in. As conceptually to me it's the same as providing a wrong password (the cookie) for authentication. 

Of course you get a session either way but the code sets the user to `User.anonymous` which makes them quite literally `unauthenticated` as we do not know who they are.

**Comment by oliverguenther** [comment]
I respect your reasoning concerning 401. My only rationale behind 403 was that the spec mandates the presence of `WWW-Authenticate` for 401 Unauthorized.

> [Sec. 12, # 2617](http://tools.ietf.org/html/rfc2617)
> The response MUST include a WWW-Authenticate header field (section 14.47) containing a challenge applicable to the requested resource.

 If we can test all clients to be compatible with custom challenges as the header value, this should pose no problem at all, as we can simply return 'session' to be conformant with the spec.

**Comment by machisuji** [comment]
I see. Yes that makes sense. Clients should have no problem with custom challenges. I'll test it with some common clients.

**Comment by oliverguenther** [comment]
Someone did that with a few common browsers (cf. 'Unknown Schemes'): http://greenbytes.de/tech/tc/httpauth/#unknown

Seems to work fine.

**Comment by machisuji** [comment]
Well, everything's fine and dandy then!

File: lib/open_project/authentication.rb [hunk 2]
```diff
@@ -61,14 +61,34 @@ def values
...
+      # Hence we rename the auth scheme in a way that makes it still obvious
+      # to developers that it is basic auth but still unknown to the browser
+      # thereby avoiding the undesired prompt.
+      def auth_scheme
+        'BasicAuth'
+      end
+
       def realm
-        'OpenProject'
+        'OpenProject API'
```
Comment (NobodysNightmare) [inline lib/open_project/authentication.rb]: What makes me nervous here is that we are in the file `lib/open_project/authentication.rb`

The file is not related to any of our APIs and still its realm now contains "API". Is this a good idea? I thought we wanted to extend the scope of warden beyond the API in the long term.

**Comment by machisuji** [comment]
@NobodysNightmare @oliverguenther Mind having another look? The `X-Authentication-Scheme` request header is now considered in the `WWW-Authenticate` response header. Also `realm` can be defined per scope.

File: lib/open_project/authentication.rb [hunk 3]
```diff
@@ -61,14 +73,68 @@ def values
...
-    module Realm
+    ##
+    # Options used in the WWW-Authenticate header returned to the user
+    # in case authentication failed (401).
+    module WWWAuthenticate
       module_function
 
-      def realm
+      def pick_auth_scheme(supported_schemes, default_scheme, request_headers = {})
+        key = request_headers.keys.find { |h| h =~ /X-Authentication-Scheme$/i }
```
Comment (oliverguenther) [inline lib/open_project/authentication.rb]: Do we really have to parse request headers ourselves? If there is no other way, `.starts_with?` is [probably faster](http://stackoverflow.com/questions/4168543/how-do-i-find-if-a-string-starts-with-an-other-string-in-ruby).

File: lib/api/utilities/grape_helper.rb [hunk 4]
```diff
@@ -48,12 +48,13 @@ def grape_error_for(env, api)
         end
       end
 
-      def error_response(rescued_error, error = nil, rescue_subclasses: nil, headers: {})
+      def error_response(rescued_error, error = nil, rescue_subclasses: nil, headers: ->() { {} })
         default_response = lambda do |e|
           representer = ::API::V3::Errors::ErrorRepresenter.new e
+          resp_headers = instance_exec &headers
```
Comment (oliverguenther) [inline lib/api/utilities/grape_helper.rb:54]: I was struggling to fully grasp the reason behind the `instance_exec`. Just to confirm: You need to evaluate `env` (from the lambda in [API::Root.auth_headers](https://github.com/machisuji/openproject/blob/feature/suppress_basic_auth_prompt/lib/api/root.rb#L147-L149)) in the context of `GrapeHelper` ?

**Comment by oliverguenther** [comment]
Looks good! Would call for merge once Hound is pleased.

File: lib/api/utilities/grape_helper.rb (hunk 4 above)
Comment (machisuji) [inline lib/api/utilities/grape_helper.rb:54]: Correct.

**Comment by ulferts** [comment]
@machisuji cukes fail (and on both DBMS so this seems to be legit)

**Comment by oliverguenther** [comment]
@ulferts Seems to be caused by one of my PRs: `Unable to find css "#main-menu" (Capybara::ElementNotFound)`. It seems #main-menu is hidden in some cases. Looking into it now.

**Comment by oliverguenther** [comment]
No, this is probably unrelated and I also can't reproduce the failing cuke locally on this PR.

**Comment by ulferts** [comment]
@machisuji can we merge in the current release/4.2 to eliminate the possibility of anything not related to this branch causing the bugs.

**Comment by machisuji** [comment]
Sure I'll do it.

**Comment by machisuji** [comment]
Will rebase if this is green then.

**Comment by machisuji** [comment]
OK this must be due to my changes then. I'm looking into it.

**Comment by machisuji** [comment]
`/api/experimental/projects/1/queries/grouped` is called without the `X-Authentication-Scheme` header. Even when I revert my change to `application_controller.rb` the cuke breaks. I wonder how this only affects my PR.

**Comment by machisuji** [comment]
But it does work on plain 4.2. Weird. Well we'll see.

**Comment by ulferts** [comment]
The cukes fail because for some reason the call to `/api/experimental/projects/1/queries/grouped` is done without proper credentials. The user then is Anonymous who is not allowed to do such a thing.

The call right before that is one to `/api/v3/queries/1/star` which is working fine. 

I haven't done the extra work of partially reverting changes made in this PR to see when it fails but a bisect tells me that 1acc69b7200f297905bc4914ff7dee674162a858 caused the problems.

**Comment by NobodysNightmare** [comment]
@ulferts @oliverguenther I admit to have lost track of this PR. Is anyone of you aware that tests are now passing? Does that mean we can merge it? (into `release/4.2` I guess)

You two were the last people talking here, that is why I am asking ^^

**Comment by machisuji** [comment]
I can re-create the PR against 4.2.

**Comment by machisuji** [comment]
Here's the [PR](https://github.com/opf/openproject/pull/3203) against `release/4.2`.

**Comment by machisuji** [comment]
Closing this in favour of the PR against `4.2`.
